cli/src/commands/fix.rs

```diff
@@ -0,0 +1,320 @@
+                        // TODO: Consider filename arguments and tool configuration instead of
+                        // passing every changed file into the tool. Otherwise, the tool has to
+                        // be modified to implement that kind of stuff.
+                        paths_to_fix.insert(repo_path.clone());
+                    }
+                }
+            }
+            Ok::<(), BackendError>(())
+        }
+        .block_on()?;
+
+        // Also fix all paths that were fixed in ancestors, so we don't lose those
+        // changes. We do this instead of rebasing onto those changes.
+        for parent_id in commit.parent_ids() {
+            if let Some(parent_tool_inputs) = commit_tool_inputs.get(parent_id) {
+                for repo_path in parent_tool_inputs.keys() {
+                    paths_to_fix.insert(repo_path.clone());
+                }
+            }
+        }
+
+        // Get this commit's file IDs for the paths to be fixed. Conflicting files will
+        // have multiple versions to be fixed, so we map paths to a set of inputs.
+        let mut tool_inputs_map: HashMap<RepoPathBuf, Vec<ToolInput>> = HashMap::new();
+        for repo_path in paths_to_fix {
+            let value = tree.path_value(&repo_path);
+            for term in value.iter().flatten() {
+                if let TreeValue::File { id, executable: _ } = term {
+                    let tool_input = ToolInput {
+                        file_id: id.clone(),
+                        repo_path: repo_path.clone(),
+                    };
+                    all_tool_inputs.insert(tool_input.clone());
+                    tool_inputs_map
+                        .entry(repo_path.clone())
+                        .or_default()
+                        .push(tool_input.clone());
+                }
+            }
+        }
+        commit_tool_inputs.insert(commit.id().clone(), tool_inputs_map);
+    }
+
+    // Run the configured tool on all of the chosen inputs.
+    // TODO: Support configuration of multiple tools and which files they affect.
+    let tool_command = command.settings().config().get_string("fix.tool-command")?;
+    let fixed_file_ids =
+        fixed_file_ids(tx.repo().store().as_ref(), &tool_command, &all_tool_inputs)?;
+
+    // Substitute the fixed file IDs into all of the affected commits. Currently,
+    // fixes cannot delete or rename files, change the executable bit, or modify
+    // other parts of the commit like the description.
+    let mut num_checked_commits = 0;
+    let mut num_fixed_commits = 0;
+    tx.mut_repo().transform_descendants(
+        command.settings(),
+        root_commits.iter().ids().cloned().collect_vec(),
+        |mut rewriter| {
+            let tool_inputs_map = commit_tool_inputs.get(rewriter.old_commit().id()).unwrap();
+            let old_tree = rewriter.old_commit().tree()?;
+            let mut tree_builder = MergedTreeBuilder::new(old_tree.id().clone());
```

Comment:
I don't think it needs to be done anytime soon, but I guess I was expecting it to be part of the initial implementation given all the other hard work here to make sure that processing is parallelized/not serialized/etc. 😉

Reply:
Ah, I see. Parallelizing against the Git backend doesn't help much AFAIK. With the Google backend, writes are still pretty fast because we have a proxying daemon process that hashes the objects, returns and ID, and then sends the objects to the server in the background. It's only reads we care about parallelizing so far. (This may change if someone else writes a backend that's slow at writing too.)